### PR TITLE
[Android][WebView] onShouldStartLoadWithRequest callback

### PR DESCRIPTION
**Pull request by pglotov**
This PR implements `onShouldStartLoadWithRequest` callback for WebView on Android. Similar to iOS approach in [PR#3643](https://github.com/facebook/react-native/pull/3643).

### CODE REVIEW

**Comment by jordansexton** [comment]
I love you so much right now.

**Comment by ide** [comment]
cc @astreet this PR introduces synchronous JS execution by allowing the UI thread to invoke JSC. CCing you since you wrote the Worker implementation. Based on my experience with multicore JSC, calling into the same JSContext from two threads is going to introduce race conditions within JS itself since JSC will try to concurrently interleave two separate JS calls.

@pglotov I don't think this approach is going to work. If the WebView method needs to be synchronous, you should queue up the JS command to be run on the JS thread and then block the UI thread until the JS command is processed.

**Comment by pglotov** [comment]
@ide Actually it synchronously calls Java from JS thread. UI thread sends an event to JS and blocks. JS processes the event and posts result into Java object, unblocking UI thread. Race conditions in Java world can be dealt with.

**Comment by dmmiller** [comment]
Can you look at how iOS does this in WebView.ios.js?  Also, would be nice if it had the same property name as in ios as well.  Have you checked how this works for back/forward and history?   Also shouldOverrideUrlLoading is not called in all instances.  For instance, not on the initial load, nor when you explicitly tell the WebView to load a URL, only on link clicks which may be good enough.

**Comment by pglotov** [comment]
@dmmiller Yes I looked at iOS `onShouldStartLoadWithRequest` before implementing this, and tried to follow that approach. No I didn't check back/forward. I need it to intercept link clicks in the webview and open them in browser instead. Makes sense to have a single property name for both platforms.

File: Libraries/Components/WebView/WebView.android.js [hunk 1]
```diff
@@ -195,6 +198,16 @@ var WebView = React.createClass({
       console.warn('WebView: `source.body` is not supported when using GET.');
     }
 
+    var onShouldOverrideUrlLoading = this.props.onShouldStartLoadWithRequest && ((event: Event) => {
```
Comment (davidaurelio) [inline Libraries/Components/WebView/WebView.android.js]: just some formal rants (aka nits): The formatting here makes this particularly hard to read. Please add some line breaks, and white space. Thank you.

File: Libraries/Components/WebView/WebView.android.js [hunk 2]
```diff
@@ -195,6 +198,16 @@ var WebView = React.createClass({
...
+      var shouldOverride = this.props.onShouldStartLoadWithRequest &&
```
Comment (davidaurelio) [inline Libraries/Components/WebView/WebView.android.js]: How can `this.props.onShouldStartLoadWithRequest` go away in the meantime?

File: Libraries/Components/WebView/WebView.android.js [hunk 3]
```diff
@@ -195,6 +198,16 @@ var WebView = React.createClass({
...
+          this.props.onShouldStartLoadWithRequest(event.nativeEvent);
+      UIManager.dispatchViewManagerCommandSync(
+        this.getWebViewHandle(),
+        UIManager.RCTWebView.Commands.shouldOverrideWithResult,
+        [shouldOverride]
+      );
+    }).bind(this);
```
Comment (davidaurelio) [inline Libraries/Components/WebView/WebView.android.js]: @eslint-bot tells the truth

**Comment by ide** [comment]
@pglotov ah sorry I must have misread what is going on

**Comment by pglotov** [comment]
@davidaurelio Thanks, made changes.

**Comment by davidaurelio** [comment]
great, thank you!

File: ReactAndroid/src/main/java/com/facebook/react/views/webview/events/TopLoadingErrorEvent.java [hunk 4]
```diff
@@ -7,7 +7,7 @@
  * of patent rights can be found in the PATENTS file in the same directory.
  */
 
-package com.facebook.catalyst.views.webview.events;
+package com.facebook.react.views.webview.events;
```
Comment (dmmiller) [inline ReactAndroid/src/main/java/com/facebook/react/views/webview/events/TopLoadingErrorEvent.java]: Thanks for cleaning this, but this is really orthagonal.  Want to put up a quick PR to fix these and then I can accept and then this one will be easier to read?

**Comment by dmmiller** [comment]
This question of whether we want to allow synchronous communication this way is pretty big.  I'd like @andreicoman11 and @astreet to weigh in on it.

File: ReactAndroid/src/main/java/com/facebook/react/views/webview/events/TopLoadingErrorEvent.java (hunk 4 above)
Comment (pglotov) [inline ReactAndroid/src/main/java/com/facebook/react/views/webview/events/TopLoadingErrorEvent.java]: Sure, will do it later today.

Comment (pglotov) [inline ReactAndroid/src/main/java/com/facebook/react/views/webview/events/TopLoadingErrorEvent.java]: Created PR for path changes [here](https://github.com/facebook/react-native/pull/6516).

File: Libraries/Components/WebView/WebView.android.js [hunk 5]
```diff
@@ -195,6 +198,15 @@ var WebView = React.createClass({
       console.warn('WebView: `source.body` is not supported when using GET.');
     }
 
+    var onShouldOverrideUrlLoading = this.props.onShouldStartLoadWithRequest
+        && ((event: Event) => {
+            var shouldOverride = this.props.onShouldStartLoadWithRequest(event.nativeEvent);
```
Comment (jordansexton) [inline Libraries/Components/WebView/WebView.android.js]: @pglotov Possible issue based on my reading of the source here. The semantics of `onShouldStartLoadWithRequest` are that if a `true` value is returned, the WebView should handle the request, whereas the semantics of `shouldOverrideUrlLoading` are the same if a `false` value is returned. It seems to me that this line should be changed to:

```javascript
var shouldOverride = !this.props.onShouldStartLoadWithRequest(event.nativeEvent);
```

Comment (pglotov) [inline Libraries/Components/WebView/WebView.android.js]: Good catch.

**Comment by astreet** [comment]
Having a mix of sync and async commands like this is a bad idea. We'll either have to destroy ordering or block on the JS queue being drained. Can you explain what you're trying to accomplish?

**Comment by pglotov** [comment]
This code allows using [`shouldOverrideUrlLoading`](http://developer.android.com/reference/android/webkit/WebViewClient.html#shouldOverrideUrlLoading(android.webkit.WebView, java.lang.String)) Android feature (renamed as iOS `onShouldStartLoadWithRequest` for uniformity sake). I need it to prevent WebView from opening links inside the view, but rather in a browser. There can be other uses for it as well.

Why ordering will be destroyed? It still calls all previous events in the current frame. However, we can put a constrain on synchronous events to be totally independent of other events and only call the sync event processor.

**Comment by ide** [comment]
@astreet there are several native APIs on both platforms -- specifically listeners like WebViewClient on Android and delegates on iOS -- that expect your code to run synchronously. I think you get this, here's a diagram for more clarity or for other people interested in learning what's going on:

```
               [      OS's native library        ]
                  +                           ^
                  |                           |
          OS calls your code    Native listener returns the  <---+
                  |             result from JS back to the OS    |
                  v                           +                  |
               [  Your native listener/delegate  ]               |
                  +                           ^                  |
                  |                           |                  |
Your native code enqueues a JS call   JS sends back result       |
                  |                           |                  |
                  v                           +                  |
               [     Your app logic in RN JS     ]               |
                                                                 |
                                                                 |
              The OS's native library expects this +-------------+
              call to return synchronously. As a
              result, we need both the call into JS
              and getting the result both to happen
              synchronously from the perspective of
              the native side.

              On iOS, the way we do this is to block
              the native UI thread until we get the
              result from the JS thread. This approach
              works with remote JS executors like the
              Chrome debugger that are inherently
              asynchronous.

```

**Comment by astreet** [comment]
Unfortunately, those APIs are going to be quite hard to expose. Is there a way we can give a synchronous answer (e.g. 'no') and then follow up once we actually get a response from JS? Can we implement our own version of the component without that synchronous requirement? Can we have JS declare to native how it should synchronously respond when the component is created? (of course, the answers to these questions differ on a case-by-case basis)

**Comment by pglotov** [comment]
RN for iOS already implemented/exposes this callback. There are many more callback Android components have that return value. Why not allow them?

**Comment by jordansexton** [comment]
I'm wondering, should this incorporate the logic iOS has for its [`react-js-navigation://` navigation](https://github.com/facebook/react-native/blob/520ad05ba07a3442f88f070458c70db5d1c74cc9/React/Views/RCTWebView.m#L187) to bypass the supplied handler? It would certainly make the API more consistent.

**Comment by astreet** [comment]
+@nicklockwood

**Comment by astreet** [comment]
>RN for iOS already implemented/exposes this callback. There are many more callbacks Android components have that return value. Why not allow them?

I didn't realize that we had already done this on iOS: I really don't think we should have. It violates our basic bridge contract, which opens up a can of worms -- what do you do if you can't get a response from JS in time? Do you just let your app hang? Do you show an error message and make the user re-request? What is the default behavior in the case you can't get a response in time, and can the app change it? Who implements that error message (framework or app)?

Can you explain how you expect to use this callback? I'm hoping we can come up with a solution that preserves most of the value of the API in the common case, but doesn't require sync messages (e.g. let JS specify a URL regex to accept/block).

**Comment by jordansexton** [comment]
In both iOS and this implementation, the thread is blocked up to 250ms to acquire a result. Based on my reading of the iOS implementation, it may resolve sooner than that. I think the Android implementation just blocks for 250ms and then checks the value, but @pglotov would know better.

Here's how I use it with this patch, but I use the same for iOS too:
```javascript
    handleShouldStartLoadWithRequest (nativeEvent) {
        var { url } = nativeEvent;
        if (/^https?:\/\/localhost:\d+(\/|$)/.test(url)) {
            return true;
        }
        else if (/^react-js-navigation:\/\//.test(url)) {
            this.handleNavigationStateChange(nativeEvent);
            return false;
        }
        else {
            if (url !== 'about:blank') {
                setTimeout(function () {
                    Linking.openURL(url);
                }, 250);
            }
            return false;
        }
    }
```

While I use regexes here, I don't think the ideal implementation would favor them. It's conceivable that other kinds of conditions might be better, and the issue is not only whether the request is allowed or denied, but also what you do decide to do with it.

**Comment by jordansexton** [comment]
I forgot to address this part of your post. The default behavior on both platforms (if the result isn't set, or if no callback is provided) is to allow the request.

**Comment by astreet** [comment]
Yeah, your example could definitely be implemented by coming up with our own API where you can declare beforehand how to handle certain URL schemes, and then use this as an async callbacks to implement your setTimeout/handleNavigationStateChange. I think that would be much safer.

>The default behavior on both platforms (if the result isn't set, or if no callback is provided) is to allow the request.

What if the app developer is trying to block navigation to a certain set of URIs? You're saying that we would just non-deterministically not respect that sometimes?

**Comment by jordansexton** [comment]
Right, so in my case, I block navigation to anything not localhost (the apps are running GCDWebServer/AsyncHTTPServer). If the thread was blocked for longer than 250ms before receiving a result, it would default open URLs it shouldn't.

Mapping URL schemes to callbacks and denying the request wouldn't be sufficient on its own, though. You may want to allow the request but also perform another action, or need other information besides the scheme as above to make a determination.

I agree that [the current implementation on iOS](https://github.com/facebook/react-native/blob/master/React/Views/RCTWebViewManager.m#L100) feels pretty sketchy, not to mention that needing to return a value from an event handler in userland code is really uncommon.

I think you suggested something like this before, but what about a flag to deny all requests, and having a mechanism to initiate desired ones from JS? This seems safer than hanging with default allow on error, though I don't know what the other complications would be.

**Comment by pglotov** [comment]
This implementation also blocks for **up to** 250ms, but usually unblocks much faster when the line
`this.mShouldOverrideUrlLoadingConditionVariable.open();` executes. Currently iOS implementation proceeds with opening the URL if it unblocks at timeout, as this Android implementation does (for consistency). We can change that and not open the URL if timeout expires. I think this would be better since if JS thread is busy this behavior would be consistent and safe.

In general what to do when timeout expires before callback finishes depends on callback, but it is not typical. At least in this case. For example we can reschedule the event. But these are corner cases, and shouldn't prevent one from allowing the feature.

In general I think since both native platforms allow sync callback, so should RN. RN should be made as flexible and generic as native (as much as possible). There probably are use cases which require functional callbacks, rather than fixed parameterized behavior.

**Comment by astreet** [comment]
>You may want to allow the request but also perform another action, or need other information besides the scheme as above to make a determination.

Since those actions would be async anyway, they don't need to be performed in a sync callback -- i.e. we could still deliver this event, but not allow responding to it in a sync manner.

>I think you suggested something like this before, but what about a flag to deny all requests, and having a mechanism to initiate desired ones from JS? This seems safer than hanging with default allow on error, though I don't know what the other complications would be.

Not sure if I understand you, but yeah I would be much happier with a way to declare to native what you do/do not want to allow beforehand so that it can respond synchronously.

>In general I think since both native platforms allow sync callback, so should RN. RN should be made as flexible and generic as native (as much as possible). There probably are use cases which require functional callbacks, rather than fixed parameterized behavior.

We should handle them on a case-by-case basis. Practically speaking, I don't know of any other places we've needed to do this so far, so allowing sync callbacks in general doesn't seem very compelling. We would also only be able to use this approach in places where we are ok with blocking the main thread for an arbitrary period of time, and each time we would have to handle the case where we don't actually get a response back from JS in a reasonable amount of time which very much depends on phone conditions. I also feel like we should just not *ever* be ok with blocking the main thread like this: it will interrupt any animations, touch events, etc and may flash a ANR (application not responding) on Android.

**Comment by pglotov** [comment]
ANR is currently set at 5 seconds, so this shouldn't be a problem.

**Comment by astreet** [comment]
Ok, ANR wouldn't be a problem then - everything else still stands though. I think my least favorite aspect is the non-determinism of behavior since you never know when you might time out.

In order to move this PR forward, can you give an example of what you're trying to achieve with this callback? @jordansexton provided an example, is that similar to your use case?

**Comment by pglotov** [comment]
In my case I simply return `false` and open url in browser with `Linking.openURL()`.

**Comment by astreet** [comment]
Then I would say we should come up with an API where JS can declare how native should synchronously respond to this event (e.g. a url regex seems to work for both of your cases) and then still send this event down as an async event that you can use to call Linking.openURL from

**Comment by pglotov** [comment]
It would be kind of .. non-native.

**Comment by astreet** [comment]
>It would be kind of .. non-native.

In what way?

**Comment by pglotov** [comment]
Not like corresponding native components APIs.

**Comment by astreet** [comment]
There are many places where we don't correspond to the native APIs as written but achieve equivalent functionality :) (see networking, layout, data storage, etc)

**Comment by ide** [comment]
Text Inputs are another API that need synchronous communication to determine what to do when the user types a keystroke. The native APIs expect a synchronous response from delegate methods like "should we accept the key the user just typed"?

Currently the way we deal with this is to let JS predeclare a policy, such as the `maxLength` prop on TextInputs. The native side implements that policy in a synchronous way. This is too limiting, though, especially when designing policies for text transformation (ex: wanting to uppercase all characters -- our current workaround is to use an all-caps font :P).

So my sense is for policies to accommodate a large number of use cases, we want to express the policies with a DSL... without actually designing our own DSL and all the problems that come with it. (Maybe you see where I'm going with this...) I propose that DSL is JavaScript: we launch a second JS VM that runs on the UI thread and is responsible for synchronously responding to delegate methods of TextInputs, WebViews, and other native APIs that expect synchronous responses.

This second JS VM would be entirely opt-in and lazily initialized the first time you use a prop that requires it. So people targeting really low-end devices could not use these props and the API would be strictly additive to give you more options if you know you're targeting iPhones or modern Android hardware.

**Comment by jordansexton** [comment]
Aside from the stylistic concerns, I don't think a regex is going to get it done in the general case and feels messy. Regular expression support is not the same in JS, Java, and ObjC. More importantly, how you handle URLs could also easily be dependent on state in the component containing the WebView or the application in general.

> a flag to deny all requests, and having a mechanism to initiate desired ones from JS

Here's some code sketching out the JS side of what I mean:

```javascript
var { onShouldStartLoadWithRequest } = this.props;
if (onShouldStartLoadWithRequest) {
    onShouldStartLoadWithRequest = this.handleShouldStartLoadWithRequest;
}

<RCTWebView
    alwaysLoadRequests={ false } 
    onShouldStartLoadWithRequest={ onShouldStartLoadWithRequest }
    refs="webView"
    ... />
```

If the value of `alwaysLoadRequests` is `true` (the default), `onShouldStartLoadWithRequest` events will be triggered, but the native WebView will always continue to load them. If `false`, it will not load requests at all, synchronously returning `false` in the delegate. Then, the `onShouldStartLoadWithRequest` event will trigger, resulting in:

```javascript
handleShouldStartLoadWithRequest (event) {
    var { onShouldStartLoadWithRequest } = this.props;
    var shouldStart;
    if (onShouldStartLoadWithRequest) {
        shouldStart = onShouldStartLoadWithRequest.call(this, event);
    }
    else {
        shouldStart = true;
    }

    if (shouldStart) {
        event.shouldStart = true;
        UIManager.dispatchViewManagerCommand(
            React.findNodeHandle(this.refs.webView),
            UIManager.RCTWebView.Commands.startRequest,
            [event]
        );
    }
}
```

So basically, all requests are stopped by default with a flag, and then there is a privileged mechanism to start requests with a flag on the event object, which can then by read in the native code examining the request. Everything over the bridge is still async, the requests get slowed down slightly, but the behavior is all deterministic, and the flow control is purely in JS and not in native or some kind of weird DSL we'd have to add for allowing requests.

**Comment by jordansexton** [comment]
Or, you know, what @ide said, because it's a bit more work, but far more robust haha.

**Comment by ide** [comment]
I think it's actually not that much work to build v1 for production. And I believe performance is acceptable on modern devices -- Android phones regularly come with 3-4 GB of RAM and iPhones all come with 1-2 GB so an extra JS VM that runs a small amount of relatively pure code (as in pure functions) should be acceptable. Or maybe not, but we could ship to production behind a flag and find out.

**Comment by astreet** [comment]
>More importantly, how you handle URLs could also easily be dependent on state in the component containing the WebView or the application in general.

Yes, this is much more of a concern here -- and this also isn't something that would be solved by running JS on the main thread (as the state data still lives on the JS thread). I'm not convinced that just having the component update the webview accept/block-list when its state updates wouldn't be sufficient here though. I suppose you could have a user interaction that causes the blocklist to update, and then another web navigation happens before that update can be processed on the main thread, but then the first user interaction would necessarily not have registered any user-visible updates (e.g. UI) by the time the second interaction happened either. This blocklist may work for the webview case, but the TextInput case requires much more flexibility (thanks for bringing that up).

Re: a separate JS VM: There's still the issue that application data (e.g. component state) lives on the JS thread, so we wouldn't be able to access that safely.

I wonder if it's feasible to extend or reimplement the native components to handle async responses (e.g. block requests in webviews/chars in TextInput and then have an automatic mechanism to re-send them once they're 'accepted' -- I imagine this could work ok for webview, but text and keyboard input is a clusterfuck even outside of RN, at least on android, so I'm not too hopeful there).

**Comment by pglotov** [comment]
> Yes, this is much more of a concern here -- and this also isn't something that would be solved by running JS on the main thread (as the state data still lives on the JS thread). I'm not convinced that just having the component update the webview accept/block-list when its state updates wouldn't be sufficient here though. I suppose you could have a user interaction that causes the blocklist to update, and then another web navigation happens before that update can be processed on the main thread, but then the first user interaction would necessarily not have registered any user-visible updates (e.g. UI) by the time the second interaction happened either. This blocklist may work for the webview case, but the TextInput case requires much more flexibility (thanks for bringing that up).

True, within one JS frame we can miss state change. Which basically means JS is super busy, oh well ..  However, I'd make it NOT load request by default. There are [Futures](http://developer.android.com/reference/java/util/concurrent/Future.html) in Android at the moment, which would still require some kind of wait mechanism for UI thread, so it's probably better to wait till Promises are available.

**Comment by mkonicek** [comment]
Is there a clear way how to proceed with this PR? @pglotov it looked like astreet wasn't happy with the changes? Do you have a plan on how to move this forward?

**Comment by pglotov** [comment]
Closing since no interest.

**Comment by jordansexton** [comment]
That's unfortunate, considering all of the problems and worse that were pointed out already exist in the iOS implementation. The proposed implementation might not be ideal, but achieving feature parity is worth something in practical terms and this PR accomplishes that.

**Comment by antihate** [comment]
Sad to see this PR closed. So, currently there is no way to prevent loading unwanted URL on Android?

**Comment by astreet** [comment]
I think it was a mistake to implement it this way on iOS, but I'm also not going to go and change it at this point. The API I think we should move to on both platforms is a way to declare a whitelist/blacklist set of URI patterns (as mentioned above).

**Comment by pglotov** [comment]
How would one specify what to do with url in case its whitelisted? Whether open it in the same webview, in a different webview, in a browser, in an app, may be add some query parameters to it, or may be make some other http request instead?

**Comment by astreet** [comment]
You would still receive the async callback in JS, e.g. as onRequestBlocked or onRequestAllowed

**Comment by pglotov** [comment]
How would you specify "whitelist/blacklist set of URI patterns"? Static regexs? What if one wants them depend on the state (on current URL)?

**Comment by astreet** [comment]
Yep, static regexes since that covers the vast majority of usecases I've heard from people. These can be updated asynchronously based on state -- though this can obviously introduce timing issues so it probably wouldn't work well for something like depending on currentURL. This isn't ideal, but we're constrained by the synchronous nature of the API. If that's something you need, let's talk specifics and I bet we can come up with a reasonable solution.

**Comment by isairz** [comment]
I glad to find what i want and sad to see this PR is closed.

**Comment by fernsheldon** [comment]
Sad to see this PR closed. This is exactly what I needed.

**Comment by vanBrunneren** [comment]
I made a thread on Stackoverflow maybe someone can answer this http://stackoverflow.com/questions/38999499/react-native-android-webview-handle-clicked-url-before-loading

**Comment by sofakeer** [comment]
2016-08-18 17:17 GMT+08:00 Pascal Brunner <notifications@github.com>:

**Comment by astreet** [comment]
I hear your pain, I'm sorry RN doesn't currently work well for this use case. To solve this, it would be helpful to understand exactly what you're trying to do to see whether it'd be helpful to add a declarative API here (or any other solution) instead of creating an one-off exception within our bridging protocol.

I had suggested above allowing the WebView component to declare a URL pattern to allow/deny and possible adding onURLAllowed/onURLDenied callbacks. Would this be helpful for what you're trying to do?

**Comment by rseemann** [comment]
@astreet I'm building an app that right now is a native wrap that bundles a web-app. All its navigation is now native, and the way I know where to go is with the `onShouldStartLoadWithRequest `. I return false when I need to navigate to another view, but need to let it load in case there is an iFrame or form submission.

The cases are not as simple as white listing, since sometimes (like iFrames loading) I want to load, but on other times very similar urls (youtube video) clicking, I want to load that on the browser.

I don't fully understand why can't we implement something similar as iOS, where the thread is blocked for 250 ms, since most of the checkings will be done below that time, otherwise the thread is released returning `true`.

**Comment by astreet** [comment]
>I don't fully understand why can't we implement something similar as iOS, where the thread is blocked for 250 ms, since most of the checkings will be done below that time, otherwise the thread is released returning true.

Are you really ok with sometimes nondeterministically loading a page you otherwise wouldn't have loaded because RN or the system is under heavy load? That doesn't seem like a reasonable API to me.

It's unfortunate that RN doesn't bridge these synchronous APIs over very well. This is something we're very aware of and hopefully can improve in the future. In your case, I think the best option is writing a custom webview view manager in Java which extends our current one and contains the logic for which pages/urls are acceptable.

**Comment by rseemann** [comment]
@astreet I guess you're right. Do you have any example of how can I do this JS callback from Java?

**Comment by Jarred-Sumner** [comment]
+1 to needing this

**Comment by yedidyak** [comment]
+1 to needing this

**Comment by mattisx** [comment]
Jup, also needing this. Doing some actions inside a WebView, and after the user is done I need to catch a button click with a dummy url and close the WebView.

**Comment by varhp** [comment]
+1 to needing this

**Comment by younthu** [comment]
Strong need this.

**Comment by evanidul** [comment]
+1

**Comment by astreet** [comment]
Like I mentioned above, I think a reasonable solution for the majority of cases would be to add a prop to WebView which is an array of regex's of urls to block synchronously. We would also want to add a new event for WebViews to be sent to JS when a URL is blocked this way so that JS can respond appropriately. Unfortunately, this isn't something we have a need for at FB so it's going to be up to an external contributor working on it, I can help review the code and offer advice.

**Comment by younthu** [comment]
@astreet , i just implemented it in our project with the regex prop way you described,  i will try to extract those code out and send a pr to the FB, hope it help others. Please help to review it then, Thanks.

**Comment by mattiashalldin** [comment]
When will onShouldStartLoadWithRequest be available for Android?

**Comment by zheng-liu-seattle** [comment]
+1 need this.

**Comment by coyer** [comment]
+1 need this

**Comment by caigehui** [comment]
+1 need this

**Comment by Livyli** [comment]
+1 need this

**Comment by ybolaris** [comment]
+1

**Comment by miguelocarvajal** [comment]
+1

**Comment by tourze** [comment]
+1

**Comment by kyangy** [comment]
+1

**Comment by stanleycyang** [comment]
+1

**Comment by 0x5e** [comment]
Without `onShouldStartLoadWithRequest` implemented on Android, the same method on iOS is useless. :(

**Comment by wesleymooiman** [comment]
+1

**Comment by MikePodgorniy** [comment]
+1

**Comment by punksta** [comment]
+1

**Comment by anuraginwadi** [comment]
+1

**Comment by shashankwadi** [comment]
+1

**Comment by pulkitsharma99** [comment]
+1

**Comment by pramodinwadi** [comment]
+1

**Comment by akhilwadi** [comment]
+1

**Comment by manjeetwadi** [comment]
+1

**Comment by simranjeet-sawhney** [comment]
+1

**Comment by rschef** [comment]
+1

**Comment by prademak** [comment]
+1

**Comment by SanlinBlackball** [comment]
+1

**Comment by braydo25** [comment]
In the meantime, there is a pending PR on the `react-native-webview-android` module with a working implementation of onShouldStartLoadWithRequest for Android here: https://github.com/lucasferreira/react-native-webview-android/pull/98
